src/graph/entities/qg_node.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018-2019 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "qg_node.h"
+#include "../graph.h"
+#include "../../util/arr.h"
+#include <assert.h>
+
+QGNode* QGNode_New(const char *label, const char *alias) {
+    QGNode *n = rm_malloc(sizeof(QGNode));
+    n->alias = NULL;
+    n->label = NULL;
+    n->labelID = GRAPH_NO_LABEL;
+    n->incoming_edges = array_new(QGEdge*, 0);
+    n->outgoing_edges = array_new(QGEdge*, 0);
+    // n->id = -1;
+    return n;
+}
+
+void QGNode_ConnectNode(QGNode* src, QGNode* dest, QGEdge* e) {
+	src->outgoing_edges = array_append(src->outgoing_edges, e);
+	dest->incoming_edges = array_append(dest->incoming_edges, e);
+}
+
+void _QGNode_RemoveEdge(QGEdge **edges, QGEdge *e) {
+	uint edge_count = array_len(edges);
+	for(uint i = 0; i < edge_count; i++) {
+		QGEdge *ie = edges[i];
+		if(e == ie) {
+			edges[i] = edges[edge_count-1];
+			array_pop(edges);
+			return;
+		}
+	}
+}
+
+void QGNode_RemoveIncomingEdge(QGNode *n, QGEdge *e) {
+	_QGNode_RemoveEdge(n->incoming_edges, e);
+}
+
+void QGNode_RemoveOutgoingEdge(QGNode *n, QGEdge *e) {
+	_QGNode_RemoveEdge(n->outgoing_edges, e);
+}
+
+int QGNode_IncomeDegree(const QGNode *n) {
+	return array_len(n->incoming_edges);
+}
+
+int QGNode_OutgoingDegree(const QGNode *n) {
+	return array_len(n->outgoing_edges);
+}
+
+int QGNode_EdgeCount(const QGNode *n) {
+	return QGNode_IncomeDegree(n) + QGNode_OutgoingDegree(n);
+}
+
+QGNode* QGNode_Clone(const QGNode *orig) {
+    QGNode *n = rm_malloc(sizeof(QGNode));
+    n->id = orig->id;
+    n->label = orig->label;
+    n->labelID = orig->labelID;
+    n->alias = orig->alias;
+    n->incoming_edges = array_new(QGEdge*, 0);
```

Comment:
I don't remember, I can only guess we don't need / shouldn't copy edges.

Reply:
We don't by design, this is only called from the `QueryGraph_Clone` routine. We clone all nodes without edges first, then introduce edges, which automatically updates the QGNode objects.